### ISSUE TITLE
Jakarta EE 10: Support Jakarta Connectors 2.1

### ISSUE THREAD

**Issue by jhanders34** [OpenLiberty/open-liberty#18748]
## Description of the high level feature, including any external spec links:  
Implement the Jakarta Connectors 2.1 specification

Milestone for 2.1 in the spec project https://github.com/eclipse-ee4j/jca-api/milestone/4 includes:

- [x] jakarta.resource.cci.MappedRecord should extend Map with generics
- [x] new 2.1 version of ra.xml deployment descriptor

In addition to the above,

 - [x] create connectors-2.1 feature, add tests for it (possibly just repeat tests), beta it and eventually GA it

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/ldn0ybw3wdr339ybowfop7dxnc3ue6w5
- FTS: https://github.com/OpenLiberty/open-liberty/issues/21602
- Beta Blog Post (if applicable): https://github.com/OpenLiberty/open-liberty/issues/22115
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/23290

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by mbroz2** [comment]
UFO Review:
Comments:
Slide 15: The approach of using connectorInboundSecurity-2.1 will be replaced by an autofeautre. Nathan to check which feature will bring in the autofeature in, as it might be appSecurity-5.0 based on the discussion.

**Comment by njr-11** [comment]
> UFO Review: Comments: Slide 15: The approach of using connectorInboundSecurity-2.1 will be replaced by an autofeautre. Nathan to check which feature will bring in the autofeature in, as it might be appSecurity-5.0 based on the discussion.

The resolution is that the appSecurity-5.0, not ssl, will remain the other triggering feature for the autofeature that replaces connectorsInboundSecurity in Connectors 2.1/EE 10.  The document slides remain as written, but I added a comment to the notes below the first mention noting that investigation had been performed and determined that API from appSecurity is needed by the inbound security bundle, confirming the design as written.

**Comment by NottyCode** [comment]
The only comment I would make is the use case shouldn't really be written as "implement the specification" vs "the use cases the specification enabled" however I'm going to give this a pass since the feature doesn't deliver a significant amount of new function overall and I understand what is added based on the UFO.

**Comment by KyleAure** [comment]
# Servicability Review

# UFO
Does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

- The connectorsInboundSecurity feature is now an auto feature. (slide 9/28)
- The jca feature has been renamed connectors (slide 21)
- Require minimum Java level of 11 (slide 24)

# Test and Demo
As part of the serviceability process, we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team). 

1. What problem paths were tested and demonstrated? 
  - The connectorsInboundSecurity feature is now an auto feature: 
    - Migration team has been notified. 
    - Documentation will be updated with this PR: https://github.com/OpenLiberty/open-liberty/pull/22617
  - The jca feature has been renamed connectors: 
    - Existing kernel warning: **CWWKF0045E: An existing feature definition, connectors-2.1, is a possible match for the jca-2.1 feature that was not found.**
  - Require a minimum Java level of 11
    - Existing kernel warning:  **CWWKF0032E: The io.openliberty.jakarta.connectors-2.1 feature requires a minimum Java runtime environment version of JavaSE 11.**

2. Who did you demo to? 
  - Greg Watts, Mark Swatosh, Nathan Mittlestat, Nathan Rauh

3. Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without the need to engage L3?
  - Yes

# SVT
SVT team is often the first team to try new features and often encounters problems setting up and using them. 
Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered. 

1. Who conducted SVT tests for this feature? 
  - None - Brian Hanczaryk did not request explicit SVT coverage.

2. Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without the need to engage L3?
  - N/A

# Service
Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.

- L2 team: WAS: L2 CET Team
- L3 team: WAS L3: JCA, Connection Management

# Metrics
Does this feature add any new metrics or emit any new JSON events? 
If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
  
- No new metrics were added. 
- Existing metrics already accounted for: https://openliberty.io/docs/latest/metrics-list.html

**Comment by tngiang73** [comment]
@KyleAure: L2 has requested STE slides for this feature.  The STE template can be found at the links below.  You can use either one to create the education.
	Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm
	Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about
Please upload the completed slides to the same 'STE Archive' BOX folder and let Thanh Giang (@tgiang) or Geeta Nadella (@gnadell) know when they're complete or provide us the Github link.   Thanks!

**Comment by malincoln** [comment]
@njr-11 is translation required for N/A?  @samwatibm fyi.
@cbridgha can you approve for externals?
Thank you.

**Comment by njr-11** [comment]
@malincoln Yes, translation is required for this as well as for every Jakarta EE 10 feature because at a minimum for every new external feature there must be a translatable description.

**Comment by malincoln** [comment]
@samwatibm since this went into an earlier beta, has the translations returned for 220011-beta to get approval?  Thanks

**Comment by samwatibm** [comment]
Yes, 220011-beta-220010 has long been translated.   I've approved on the assumption we don't need to wait for 220012 translations

**Comment by chirp1** [comment]
- From Marylin, no ID is required for this epic: https://github.ibm.com/websphere/twas-docs/issues/5815
- From Kyle the only doc updates are not specific to anything that the documentation team would document: https://github.com/OpenLiberty/open-liberty/pull/22617
- I removed the ID required label
- Approving.
